### PR TITLE
Upgrade to WASM and Rust Node.js addon (N-API).

### PR DESCRIPTION
**Pull request by fanatid**
WIP... huge upgrade: move from [C++ Node.js addon](https://nodejs.org/api/addons.html) ([NAN](https://github.com/nodejs/nan)) and JS ([elliptic](https://www.npmjs.com/package/elliptic)) to [WebAssembly](https://webassembly.org/) and [Rust](https://www.rust-lang.org/) Node.js addon (~probably [napi](https://crates.io/crates/napi), or maybe [neon](https://crates.io/crates/neon)?~ [napi](https://crates.io/crates/napi)).

At the moment of creating pull request:
- Remove C++ addon code.
- Remove CI ([Travis CI](https://travis-ci.org/) need to be replaced with [GitHub Actions](https://github.com/features/actions)).
- Add initial WASM code (no [wasm-pack](https://rustwasm.github.io/wasm-pack/) and [wasm-bindgen](https://crates.io/crates/wasm-bindgen), because we will no use allocations and code is pretty simple).
- Add [Prettier](https://prettier.io/) as formatted for JS.
- Set minimal Node.js version to v12. v10 will be end-of-life after 2021-04-30. See https://github.com/nodejs/LTS.

Next steps for moving pull request from the draft:
- [x] Update tests (no [Buffer](https://nodejs.org/api/buffer.html), we use [Uint8Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) directly). https://github.com/fanatid/tiny-secp256k1/pull/1
- [x] ~Remove [secp256k1-sys](https://crates.io/crates/secp256k1-sys) crate. We can use [bitcoin-core/secp256k1](https://github.com/bitcoin-core/secp256k1) directly with minimum additional things. At creation moment optimized release WASM file is 400KB, what is insane.~ Investigate how to reduce size with `secp256k1-sys`. https://github.com/fanatid/tiny-secp256k1/pull/3
- [x] Attempt to write code with [no_std](https://docs.rust-embedded.org/book/intro/no-std.html). https://github.com/fanatid/tiny-secp256k1/pull/3
- [x] Add code coverage. https://github.com/fanatid/tiny-secp256k1/pull/13
- [x] ~Add tests to our `secp256k1` FFI crate.~
- [x] Add benchmarks. https://github.com/fanatid/tiny-secp256k1/pull/7 https://github.com/fanatid/tiny-secp256k1/pull/8
- [x] Implement Node.js addon. https://github.com/fanatid/tiny-secp256k1/pull/5
- [x] Reduce addon size ~(`profile.release` in `Cargo.toml`, `#![no_std]`)~, or use [optionalDependencies](https://napi.rs/introduction/release#3-the-native-addon-for-different-platforms-is-distributed-through-different-npm-packages)? https://github.com/fanatid/tiny-secp256k1/pull/8
- [x] ~Add [prebuildify](https://www.npmjs.com/package/prebuildify) for Node.js addon.~
- [x] Change [Node.js require](https://nodejs.org/en/knowledge/getting-started/what-is-require/) to [ES6 import](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import). ~Node.js (at least v12.0.0) does not support modules great, so `import` will be good for browsers, while for Node.js we will need to convert our code to `require`.~ https://github.com/fanatid/tiny-secp256k1/pull/4/
- [x] Use [TypeScript](https://www.typescriptlang.org/). https://github.com/fanatid/tiny-secp256k1/pull/9
- [x] Add usage examples. Node.js / browser ([webpack](https://webpack.js.org/)) / simple [React](https://reactjs.org/) app. https://github.com/fanatid/tiny-secp256k1/pull/11
- [x] Add CI ([GitHub Actions](https://github.com/features/actions)). https://github.com/fanatid/tiny-secp256k1/pull/14
- [x] Reduce stack size for wasm. Allow set custom stack size: https://github.com/fanatid/tiny-secp256k1/pull/6
- [x] Add Dockerfile for building / development. https://github.com/fanatid/tiny-secp256k1/pull/10

<hr>
I hope that I can handle all this :sweat_smile:

### CODE REVIEW

File: lib/validate.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  94,
+  140,
+  208,
+  54,
+  65,
+  65,
+]);
+
+function isUint8Array(value) {
+  return value instanceof Uint8Array;
```
Comment (junderw) [inline lib/validate.js:56]: iirc there was an issue in another library where some environments this doesn't work well.

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  try {
+    PUBLIC_KEY_INPUT.set(p);
+    return wasm.isPoint(p.length) === 1;
+  } finally {
+    PUBLIC_KEY_INPUT.fill(0);
+  }
+}
+
+function isPointCompressed(p) {
+  return isPoint(p) && p.length === validate.PUBLIC_KEY_COMPRESSED_SIZE;
```
Comment (junderw) [inline lib/index.js:62]: Size check should come first, it makes it much faster for the case when you pass uncompressed.

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+function isPrivate(x) {
+  return validate.isPrivate(x);
+}
+
+function pointAdd(pA, pB, compressed) {
```
Comment (junderw) [inline lib/index.js:69]: Will pA be modified in place?

**Comment by junderw** [comment]
> At creation moment optimized release WASM file is ~400KB, what is insane.

My crappy wasm-pack WASM which did so many bad inefficient rust things was 153KB... but when it is gzipped it is reduced by size 65%.

Maybe some of the optimization tools I used (like wee_alloc and the WASM strippers I used in the docker command etc.) will help. I see you are already using wasm-opt.

`wasm-strip` https://github.com/WebAssembly/wabt
`wee_alloc` https://github.com/rustwasm/wee_alloc

**Comment by junderw** [comment]
This PR is historic and godlike.

Moving JS libraries to depend on libsecp256k1 via a slim interface built from a single rust code base compiled into WASM and NAPI/NEON bindings is a huge leap towards having a "less insecure" JS Bitcoin ecosystem.

File: lib/validate.js (hunk 1 above)
Comment (fanatid) [reply lib/validate.js:56]: I remember only this issue -- https://github.com/cryptocoinjs/secp256k1-node/issues/175 :thinking: I think that `value` can be anything while it's correct work with [Uint8Array.prototype.set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/set) (we copy input value data to WASM input buffer with `set`).

File: lib/index.js (hunk 2 above)
Comment (fanatid) [reply lib/index.js:62]: We should check that `p` Array-like type first :) but yes, check length first will be more performant. Will fix.

File: lib/index.js (hunk 3 above)
Comment (fanatid) [reply lib/index.js:69]: No, we can not modify JS values from WASM while they do not refer to WASM memory. `pA` copied to input WASM buffer, and if addition successful we return new `Uint8Array`.

**Comment by fanatid** [comment]
> My crappy wasm-pack WASM which did so many bad inefficient rust things was 153KB

`wasm-bindgen` which `wasm-pack` uses under the hood is an awesome project. Currently, 400KB because std allocator (but I hope that I'll able to exclude it completely), all functions from `bitcoin-core/secp256k1` exported, etc...
Just checked [junderw/tiny-secp256k1-wasm](https://github.com/junderw/tiny-secp256k1-wasm) master and I get 289KB (91KB after gzip) for `wasm-bindgen` + `wasm-strip` + `wasm-opt`.
I'm not sure what [secp256k1.wasm](https://github.com/bitauth/libauth/blob/master/src/lib/bin/secp256k1/secp256k1.wasm) in [bitauth/libauth](https://github.com/bitauth/libauth) include, but size is 176KB. I think we should. This looks like a good size.

**Comment by fanatid** [comment]
Finally, first benchmark! Wasm vs js: 2x for sign and pk creation, 4x for verify, etc.

<details><summary>full output</summary>

```
====================================================================================================
Benchmarking function: isPoint
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 2.57 us/op (389590.66 op/s), ±3.07 %
tiny-secp256k1 (wasm): 24.49 us/op (40839.50 op/s), ±2.81 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 2.86 us/op (349568.24 op/s), ±3.05 %
tiny-secp256k1@1.1.6 (elliptic): 91.66 us/op (10910.32 op/s), ±2.66 %
secp256k1@4.0.2 (C++ addon, N-API): 3.23 us/op (310048.31 op/s), ±3.01 %
secp256k1@4.0.2 (elliptic): 92.26 us/op (10839.08 op/s), ±2.82 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1 (rust addon)
====================================================================================================
Benchmarking function: isPrivate
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 0.02 us/op (40896080.36 op/s), ±5.88 %
tiny-secp256k1 (wasm): 0.03 us/op (38838795.43 op/s), ±5.69 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 0.18 us/op (5475140.47 op/s), ±3.70 %
tiny-secp256k1@1.1.6 (elliptic): 0.35 us/op (2889282.10 op/s), ±3.14 %
secp256k1@4.0.2 (C++ addon, N-API): 0.40 us/op (2524460.19 op/s), ±6.11 %
secp256k1@4.0.2 (elliptic): 0.28 us/op (3559401.45 op/s), ±6.42 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1 (rust addon)
====================================================================================================
Benchmarking function: pointAdd
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 17.37 us/op (57559.79 op/s), ±8.01 %
tiny-secp256k1 (wasm): 150.92 us/op (6625.84 op/s), ±2.95 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 17.66 us/op (56622.59 op/s), ±3.36 %
tiny-secp256k1@1.1.6 (elliptic): 878.39 us/op (1138.45 op/s), ±5.72 %
secp256k1@4.0.2 (C++ addon, N-API): 19.02 us/op (52575.62 op/s), ±4.10 %
secp256k1@4.0.2 (elliptic): 474.96 us/op (2105.44 op/s), ±2.78 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1 (rust addon)
====================================================================================================
Benchmarking function: pointAddScalar
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 31.55 us/op (31694.21 op/s), ±5.41 %
tiny-secp256k1 (wasm): 286.57 us/op (3489.51 op/s), ±3.17 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 30.92 us/op (32338.97 op/s), ±3.70 %
tiny-secp256k1@1.1.6 (elliptic): 898.08 us/op (1113.49 op/s), ±3.79 %
secp256k1@4.0.2 (C++ addon, N-API): 32.63 us/op (30643.85 op/s), ±4.33 %
secp256k1@4.0.2 (elliptic): 701.18 us/op (1426.17 op/s), ±3.62 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
Benchmarking function: pointCompress
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 4.97 us/op (201078.76 op/s), ±23.30 %
tiny-secp256k1 (wasm): 25.05 us/op (39913.99 op/s), ±3.14 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 3.81 us/op (262203.73 op/s), ±6.32 %
tiny-secp256k1@1.1.6 (elliptic): 192.66 us/op (5190.39 op/s), ±3.08 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
Benchmarking function: pointFromScalar
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 31.17 us/op (32081.07 op/s), ±6.49 %
tiny-secp256k1 (wasm): 280.10 us/op (3570.21 op/s), ±4.32 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 30.12 us/op (33203.48 op/s), ±5.82 %
tiny-secp256k1@1.1.6 (elliptic): 542.96 us/op (1841.75 op/s), ±3.83 %
secp256k1@4.0.2 (C++ addon, N-API): 30.96 us/op (32298.58 op/s), ±3.76 %
secp256k1@4.0.2 (elliptic): 543.14 us/op (1841.16 op/s), ±3.76 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
Benchmarking function: pointMultiply
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 16.20 us/op (61742.44 op/s), ±7.38 %
tiny-secp256k1 (wasm): 139.97 us/op (7144.16 op/s), ±3.46 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 16.05 us/op (62298.03 op/s), ±4.11 %
tiny-secp256k1@1.1.6 (elliptic): 456.16 us/op (2192.23 op/s), ±5.16 %
secp256k1@4.0.2 (C++ addon, N-API): 17.13 us/op (58364.12 op/s), ±6.26 %
secp256k1@4.0.2 (elliptic): 261.01 us/op (3831.28 op/s), ±4.40 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
Benchmarking function: privateAdd
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 2.41 us/op (415199.33 op/s), ±12.69 %
tiny-secp256k1 (wasm): 0.46 us/op (2166983.92 op/s), ±3.50 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 1.73 us/op (577760.82 op/s), ±6.60 %
tiny-secp256k1@1.1.6 (elliptic): 2.50 us/op (400189.68 op/s), ±3.38 %
secp256k1@4.0.2 (C++ addon, N-API): 2.76 us/op (362045.41 op/s), ±29.20 %
secp256k1@4.0.2 (elliptic): 1.74 us/op (574956.11 op/s), ±3.25 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1 (wasm)
====================================================================================================
Benchmarking function: privateSub
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 4.31 us/op (232056.60 op/s), ±27.53 %
tiny-secp256k1 (wasm): 0.62 us/op (1606071.15 op/s), ±3.82 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 1.63 us/op (612607.01 op/s), ±10.57 %
tiny-secp256k1@1.1.6 (elliptic): 2.54 us/op (393154.52 op/s), ±3.51 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1 (wasm)
====================================================================================================
Benchmarking function: sign
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 51.71 us/op (19340.42 op/s), ±2.92 %
tiny-secp256k1 (wasm): 469.03 us/op (2132.07 op/s), ±1.62 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 49.33 us/op (20272.72 op/s), ±3.16 %
tiny-secp256k1@1.1.6 (elliptic): 913.58 us/op (1094.60 op/s), ±0.64 %
secp256k1@4.0.2 (C++ addon, N-API): 50.89 us/op (19652.12 op/s), ±2.41 %
secp256k1@4.0.2 (elliptic): 949.71 us/op (1052.95 op/s), ±0.47 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
Benchmarking function: verify
----------------------------------------------------------------------------------------------------
tiny-secp256k1 (rust addon): 60.34 us/op (16572.31 op/s), ±2.62 %
tiny-secp256k1 (wasm): 605.01 us/op (1652.87 op/s), ±0.41 %
tiny-secp256k1@1.1.6 (C++ addon, NAN/V8): 58.61 us/op (17060.88 op/s), ±2.64 %
tiny-secp256k1@1.1.6 (elliptic): 2576.87 us/op (388.07 op/s), ±0.02 %
secp256k1@4.0.2 (C++ addon, N-API): 60.74 us/op (16464.75 op/s), ±2.48 %
secp256k1@4.0.2 (elliptic): 2290.39 us/op (436.61 op/s), ±0.21 %
----------------------------------------------------------------------------------------------------
Fastest: tiny-secp256k1@1.1.6 (C++ addon, NAN/V8)
====================================================================================================
```
</details>

File: lib/validate.js (hunk 1 above)
Comment (fanatid) [reply lib/validate.js]: Do you think it worth to change to something like:
```js
function itemsInRange(value) {
  for (let i = 0; i < value.length; ++i) {
    if (value[i] < 0 || value[i] > 0xff) return false;
  }
  return true;
}

function isUint8ArrayLike(value, length) {
  return value instanceof Uint8Array || (
    value !== undefined && value !== null && value.length === length && itemsInRange(value)
  )
}
```

**Comment by fanatid** [comment]
Need few more things:
- [ ] Check cross compilation for apple m1
- [ ] Investigate wasm size (CI produce 90KB, locally I get 190KB)
- [ ] Add workflow to publish `examples/react-app` to `gh-pages`

but this should not prevent review.

Hope that review will not take a long time @junderw :laughing:

**Comment by junderw** [comment]
I am still in the middle of review.

One quick fix I would like to see:

- Since we will bump major anyways, we should join sign and signWithEntropy. The Extra Data should just be an optional argument for sign... so basically remove sign and replace it with signWithEntropy.

**Comment by fanatid** [comment]
Changed in https://github.com/fanatid/tiny-secp256k1/pull/15.

**Comment by fanatid** [comment]
Changed to ES6 import only in https://github.com/fanatid/tiny-secp256k1/pull/16

**Comment by junderw** [comment]
tACK 364b416568e325510c1899d6661510ba1ccffcd1

**Comment by junderw** [comment]
Leftover tasks:

1. README still says we are using C library. Which is true... but it should be more descriptive of how we use C -> Rust -> N-API / WASM.
2. For releasing v2.0.0 I want to have some sort of PR where we can agree on a hash for the packaged WASM.
  - I will make a PR with v2.0.0 commit.
  - We will both build and create npm package with `npm pack` and try to get the same hash, or figure out why diff is occurring...
  - If not possible we should at least try to get WASM to same hash together.
  - Once we get the same hash and share it on the pull request, I will publish 2.0.0

**Comment by fanatid** [comment]
Moved to #55

**Pull request closed without merging**